### ISSUE TITLE
[JENKINS-42637] Jenkins stuck UI does not open at all

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-06-a2#10403]
<p>Jenkins 2.49 Don't know what is actually the problem. Memory dump available for in-house usage via Dropbox etc. if needed. Please send me an email.</p>

<p> </p>

<p>Seems to be related to <a href="https://issues.apache.org/jira/browse/GROOVY-8067" class="external-link" target="_blank" rel="nofollow noopener">https://issues.apache.org/jira/browse/GROOVY-8067</a> Groovy bug. Don't know when Jenkins groovy will be update from 2.4.8 to 2.4.9 version .</p>

**Attachments:**

- [jenkins_stuck.log](https://issues.jenkins.io/secure/attachment/36366/jenkins_stuck.log)
- [jenkins_stuck2.log](https://issues.jenkins.io/secure/attachment/36376/jenkins_stuck2.log)

---
<details><summary><i>Originally reported by <a title="Heikki Simperi" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=heikkisi">heikkisi</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-42637" target="_blank">Jenkins stuck UI does not open at all</a></i></summary>
<i><ul>
<li><b>reported by</b>: heikkisi
<li><b>status</b>: Resolved
<li><b>priority</b>: Critical
<li><b>resolution</b>: Cannot Reproduce
<li><b>resolved</b>: 2017-05-08T16:30:09+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 8
<li><b>imported</b>: 2025-11-06
</ul></i>
</details>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=heikkisi">heikkisi</a>:</i>
<p>Added second thread dump. The problem occures wery often, few times per day.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=heikkisi">heikkisi</a>:</i>
<p>I'm not sure if the problem is actual Critical level issue.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>:</i>
<p>From what I see it happens in the Groovy class management logic. So my guess is that it is just another follow-up to the Groovy update to 2.4.8 in 2.47 (for <a href="https://issues.jenkins.io/browse/JENKINS-33358" title="Groovy and PermGen memory leak" class="issue-link" data-issue-key="JENKINS-33358"><del>JENKINS-33358</del></a>).</p>

<p>The pattern is close to <a href="https://issues.jenkins.io/browse/JENKINS-42189" title="Groovy 2.4.8 interoperability issues" class="issue-link" data-issue-key="JENKINS-42189"><del>JENKINS-42189</del></a></p>

<p>CC <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=daspilker" class="user-hover" rel="daspilker">daspilker</a> and <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a></p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<p>No idea offhand. Reproducible from scratch?</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=heikkisi">heikkisi</a>:</i>
<p>Our Lead CI developer Timo thinks that Groovy 2.49 will help because of this fix <a href="https://issues.apache.org/jira/browse/GROOVY-8067" class="external-link" target="_blank" rel="nofollow noopener">https://issues.apache.org/jira/browse/GROOVY-8067</a> </p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=heikkisi">heikkisi</a>:</i>
<p>This was found from thread dump:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java"><span class="code-quote">"Executor #37 <span class="code-keyword">for</span> xxx.xxxx.xxx mini : executing xxxx Healthcheck #4921725"</span> Id=580323 Group=main BLOCKED on org.codehaus.groovy.util.ManagedLinkedList@3fe823dd owned by <span class="code-quote">"Executor #28 <span class="code-keyword">for</span> xxxx.xxxx.xxx mini : executing ETJ2 Healthcheck #4921723"</span> Id=580319
                             at org.codehaus.groovy.reflection.ClassInfo$GlobalClassSet.add(ClassInfo.java:477)
                             -  blocked on org.codehaus.groovy.util.ManagedLinkedList@3fe823dd
                             at org.codehaus.groovy.reflection.ClassInfo$1.computeValue(ClassInfo.java:83)
                             at org.codehaus.groovy.reflection.ClassInfo$1.computeValue(ClassInfo.java:79)
                             at org.codehaus.groovy.reflection.GroovyClassValuePreJava7$EntryWithValue.<init>(GroovyClassValuePreJava7.java:37)
                             at org.codehaus.groovy.reflection.GroovyClassValuePreJava7$GroovyClassValuePreJava7Segment.createEntry(GroovyClassValuePreJava7.java:64)
                             at org.codehaus.groovy.reflection.GroovyClassValuePreJava7$GroovyClassValuePreJava7Segment.createEntry(GroovyClassValuePreJava7.java:55)
                             at org.codehaus.groovy.util.AbstractConcurrentMap$Segment.put(AbstractConcurrentMap.java:157)
                             at org.codehaus.groovy.util.AbstractConcurrentMap$Segment.getOrPut(AbstractConcurrentMap.java:100)
                             at org.codehaus.groovy.util.AbstractConcurrentMap.getOrPut(AbstractConcurrentMap.java:38)
                             at org.codehaus.groovy.reflection.GroovyClassValuePreJava7.get(GroovyClassValuePreJava7.java:94)
                             at org.codehaus.groovy.reflection.ClassInfo.getClassInfo(ClassInfo.java:144)
                             at org.codehaus.groovy.runtime.metaclass.MetaClassRegistryImpl.getMetaClass(MetaClassRegistryImpl.java:258)
                             at org.codehaus.groovy.runtime.InvokerHelper.getMetaClass(InvokerHelper.java:883)
                             at groovy.lang.GroovyObjectSupport.<init>(GroovyObjectSupport.java:34)
                             at groovy.lang.Script.<init>(Script.java:42)
                             at Script1.<init>(Script1.groovy)
                             at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
                             at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
                             at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
                             at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
                             at org.codehaus.groovy.runtime.InvokerHelper.createScript(InvokerHelper.java:431)
                             at groovy.lang.GroovyShell.parse(GroovyShell.java:700)
                             at groovy.lang.GroovyShell.evaluate(GroovyShell.java:584)
                             at groovy.lang.GroovyShell.evaluate(GroovyShell.java:623)
                             at groovy.lang.GroovyShell.evaluate(GroovyShell.java:594)
                             at org.jenkinsci.plugins.envinject.service.EnvInjectEnvVars.executeAndGetMapGroovyScript(EnvInjectEnvVars.java:153)
                             at org.jenkinsci.plugins.envinject.EnvInjectBuildWrapper.setUp(EnvInjectBuildWrapper.java:101)
                             at hudson.model.Build$BuildExecution.doRun(Build.java:157)
                             at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:534)
                             at hudson.model.Run.execute(Run.java:1728)
                             at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
                             at hudson.model.ResourceController.execute(ResourceController.java:98)
                             at hudson.model.Executor.run(Executor.java:405)</pre>
</div></div>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svanoort">svanoort</a>:</i>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=heikkisi" class="user-hover" rel="heikkisi">heikkisi</a>  We understand this is a critical issue for you, and in order to solve it, can you provide an isolated example that will reproduce this without depending on your infra?    </p>

<p>I think the reason we're hesitant to update Groovy more is that every update seems to introduce a new issue of this sort, which generally requires a significant time investment from someone deeply specialized in this area. </p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=grayaii">grayaii</a>:</i>
<p> </p>

<p>Not sure if this will help, but in our /etc/sysconfig/jenkins you can added:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">JENKINS_JAVA_OPTIONS=<span class="code-quote">"-Djava.awt.headless=<span class="code-keyword">true</span> -Dgroovy.use.classvalue=<span class="code-keyword">true</span> -Xms4096m -Xmx4096m -XX:MaxPermSize=1024m -XX:+CMSClassUnloadingEnabled -XX:+UseConcMarkSweepGC -Dhudson.model.ParametersAction.keepUndefinedParameters=<span class="code-keyword">true</span>"</span>
</pre>
</div></div>
<p>We also run some crazy groovy code in some of jenkins jobs and we tend to run out of memory too, so we've installed this plugin to help us track java resources:</p>

<p> </p>

<p><a href="https://wiki.jenkins-ci.org/display/JENKINS/Monitoring" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.jenkins-ci.org/display/JENKINS/Monitoring</a></p>

<p>And finally, we periodically (once per hour i think) run this groovy script to clean things up:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java"><span class="code-keyword">import</span> net.bull.javamelody.*;

before = <span class="code-object">Runtime</span>.getRuntime().totalMemory() - <span class="code-object">Runtime</span>.getRuntime().freeMemory();
<span class="code-object">System</span>.gc(); 
after = <span class="code-object">Runtime</span>.getRuntime().totalMemory() - <span class="code-object">Runtime</span>.getRuntime().freeMemory();
println I18N.getFormattedString(<span class="code-quote">"ramasse_miette_execute"</span>, <span class="code-object">Math</span>.round((before - after) / 1024));

</pre>
</div></div>
<p>And we launch it like this:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">java -jar jenkins-cli.jar -noCertificateCheck -i id_rsa -s JENKINS_URL groovy my_groovy_script.groovy'
</pre>
</div></div>
<p>You can probably put that groovy script as a jenkins job have it run periodically.</p>

<p>Ever since we we run that System.gc() command, we've never run out of memory.  We run 100s of jobs a day a AWS t2.medium without any downtime for months at a time.  Before I did these things, we were running on a huge instance and we had to restart it every week or so.</p>

<p>Hope this helps!</p>

<p> </p>
